File: src/amrte/web/administration.py
from __future__ import annotations

from typing import Any

from amrte.core.constants import AMRTE_VERSION
from amrte.operations.runtime import (
    DATASET_ID,
    EXPERIMENT_ID,
)


VERSION = AMRTE_VERSION


MANDATORY_SERVICES = (
    "configuration",
    "state",
    "clock",
    "audit",
    "observability",
    "execution",
    "health",
    "random",
)


def _name(value: Any) -> str:
    if hasattr(value, "name"):
        return str(value.name)

    return str(value)


def _configuration_identity(
    configuration: Any,
) -> dict[str, object]:
    return {
        "snapshot_id": (
            configuration.snapshot_id
        ),
        "configuration_hash": (
            configuration.configuration_hash
        ),
    }


def _registered_services(
    engine: Any,
) -> list[dict[str, object]]:
    projected: list[
        dict[str, object]
    ] = []

    for name in MANDATORY_SERVICES:
        service = engine.registry.require(
            name
        )

        projected.append(
            {
                "name": name,
                "type": type(
                    service
                ).__name__,
            }
        )

    return projected


def _persistence_projection(
    runtime: Any,
) -> dict[str, object]:
    repository = runtime.repository

    if repository is None:
        return {
            "repository": None,
            "chain_valid": False,
            "mutation_allowed": False,
        }

    chain = repository.validate_chain()

    return {
        "repository": type(
            repository
        ).__name__,
        "chain_valid": bool(
            chain.success
        ),
        "mutation_allowed": False,
    }


def build_administration_summary(
    runtime: Any,
) -> dict[str, object]:
    engine = runtime.engine

    if engine is None:
        raise RuntimeError(
            "runtime engine is unavailable"
        )

    configuration = engine.configuration

    execution = engine.registry.require(
        "execution"
    )

    return {
        "mode": "READ_ONLY",
        "research_only": True,
        "system_identity": {
            "version": VERSION,
            "environment": (
                engine.environment.name
            ),
            "experiment_id": (
                EXPERIMENT_ID
            ),
            "dataset_id": (
                DATASET_ID
            ),
        },
        "runtime": {
            "state": runtime.state,
            "environment": (
                engine.environment.name
            ),
        },
        "configuration": (
            _configuration_identity(
                configuration
            )
        ),
        "registered_services": (
            _registered_services(
                engine
            )
        ),
        "persistence": (
            _persistence_projection(
                runtime
            )
        ),
        "permissions": {
            "administration_rbac": False,
            "claims_authority": False,
            "invented_permissions": False,
        },
        "capabilities": {
            "view_system_identity": True,
            "view_services": True,
            "view_safeguards": True,
            "mutate_configuration": False,
            "register_services": False,
            "mutate_persistence": False,
            "run_diagnostics": False,
            "restore_state": False,
            "enable_execution": False,
        },
        "execution_boundary": {
            "status": "PROHIBITED",
            "provider": type(
                execution
            ).__name__,
            "financial_execution": False,
            "broker_connectivity": False,
            "account_connectivity": False,
        },
    }

File: src/amrte/web/static/administration.js
(() => {
    "use strict";

    const endpoint = "/api/v1/administration";

    const byId = (id) => document.getElementById(id);

    const text = (id, value) => {
        const node = byId(id);

        if (!node) {
            return;
        }

        node.textContent =
            value === null || value === undefined
                ? "UNAVAILABLE"
                : String(value);
    };

    const yesNo = (value) =>
        value ? "YES" : "NO";

    const allowed = (value) =>
        value ? "ALLOWED" : "PROHIBITED";

    const renderServices = (services) => {
        const container = byId(
            "registered-services"
        );

        if (!container) {
            return;
        }

        container.replaceChildren();

        for (const service of services || []) {
            const item =
                document.createElement("div");

            item.className = "detail-item";

            const label =
                document.createElement("span");

            label.textContent = service.name;

            const value =
                document.createElement("strong");

            value.textContent = service.type;

            item.append(
                label,
                value
            );

            container.appendChild(item);
        }

        text(
            "service-count",
            `${(services || []).length} SERVICES`
        );
    };

    const renderCapabilities = (
        capabilities
    ) => {
        const container = byId(
            "capability-list"
        );

        if (!container) {
            return;
        }

        container.replaceChildren();

        for (
            const [name, value]
            of Object.entries(
                capabilities || {}
            )
        ) {
            const item =
                document.createElement("div");

            item.className = "detail-item";

            const label =
                document.createElement("span");

            label.textContent =
                name
                    .replaceAll("_", " ")
                    .toUpperCase();

            const result =
                document.createElement("strong");

            result.textContent =
                allowed(Boolean(value));

            item.append(
                label,
                result
            );

            container.appendChild(item);
        }
    };

    const render = (body) => {
        text(
            "system-version",
            body.system_identity?.version
        );

        text(
            "system-environment",
            body.system_identity?.environment
        );

        text(
            "runtime-state",
            body.runtime?.state
        );

        text(
            "execution-status",
            body.execution_boundary?.status
        );

        text(
            "configuration-snapshot",
            body.configuration?.snapshot_id
        );

        text(
            "configuration-hash",
            body.configuration?.configuration_hash
        );

        text(
            "experiment-id",
            body.system_identity?.experiment_id
        );

        text(
            "dataset-id",
            body.system_identity?.dataset_id
        );

        text(
            "repository-type",
            body.persistence?.repository
        );

        text(
            "chain-valid",
            yesNo(
                Boolean(
                    body.persistence?.chain_valid
                )
            )
        );

        text(
            "persistence-mutation",
            allowed(
                Boolean(
                    body.persistence?.mutation_allowed
                )
            )
        );

        text(
            "administration-rbac",
            yesNo(
                Boolean(
                    body.permissions?.administration_rbac
                )
            )
        );

        text(
            "claims-authority",
            yesNo(
                Boolean(
                    body.permissions?.claims_authority
                )
            )
        );

        text(
            "invented-permissions",
            yesNo(
                Boolean(
                    body.permissions?.invented_permissions
                )
            )
        );

        text(
            "execution-provider",
            body.execution_boundary?.provider
        );

        text(
            "financial-execution",
            allowed(
                Boolean(
                    body.execution_boundary
                        ?.financial_execution
                )
            )
        );

        text(
            "broker-connectivity",
            allowed(
                Boolean(
                    body.execution_boundary
                        ?.broker_connectivity
                )
            )
        );

        text(
            "account-connectivity",
            allowed(
                Boolean(
                    body.execution_boundary
                        ?.account_connectivity
                )
            )
        );

        renderServices(
            body.registered_services
        );

        renderCapabilities(
            body.capabilities
        );
    };

    const load = async () => {
        try {
            const response = await fetch(
                endpoint,
                {
                    method: "GET",
                    headers: {
                        "Accept":
                            "application/json",
                    },
                }
            );

            if (!response.ok) {
                throw new Error(
                    `Administration request failed: ${
                        response.status
                    }`
                );
            }

            const body =
                await response.json();

            render(body);
        } catch (error) {
            const banner = byId(
                "administration-error"
            );

            if (banner) {
                banner.hidden = false;
            }

            console.error(
                "Administration projection unavailable.",
                error
            );
        }
    };

    load();
})();
